src/Expansion.php

```diff
@@ -332,6 +330,49 @@ private function lookupAstOptional(Ast $label, Ast $context) : Ast {
         return $result;
     }
 
+    private function doExpanderCall(Ast $expanderAst, $cg) {
+        $function = $cg->this->compileCallable('\Yay\Dsl\Expanders\\', $expanderAst->{'* expander'}, self::E_BAD_EXPANDER);
+        $expander = new \ReflectionFunction($function);
+        if ($expander->getParameters()) {
+            if ($expander->getParameters()[0]->hasType() && Ast::class === $expander->getParameters()[0]->getType()->getName()) {
+                return $this->doAstExpanderCall($expander, $expanderAst, $cg);
+            }
+
+            return $this->doTokenStreamExpanderCall($expander, $expanderAst, $cg);
+        }
+
+        return $expander->invoke(null, $cg->engine);
+    }
+
+    private function doTokenStreamExpanderCall(\ReflectionFunction $expander, Ast $expanderAst, $cg): TokenStream {
+        return $expander->invoke(
+            $cg->this->mutate(TokenStream::fromSlice($expanderAst->{'* args'}->tokens()), $cg->context, $cg->engine),
+            $cg->engine
+        );
+    }
+
+    private function doAstExpanderCall(\ReflectionFunction $expander, Ast $expanderAst, $cg): Ast {
```

Comment:
One thing I've found is that it's perfectly reasonable to want to take in an `AST` but give out a `TokenStream`; especially since that's what you're converting the resulting `AST` to. Removing this return type hint would help.

Reply:
Agreed, `fn(Ast): TokenStream` seems reasonable. It's  `fn(TokenStream): Ast` that makes not much sense for an expander.

I've removed the return type :+1: